### ISSUE TITLE
Use specialized exceptions instead of DocumentClientException everywhere

### ISSUE THREAD

**Issue by RobertoPrevato** [Azure/azure-documentdb-dotnet#348]
Hi,
This is just a recommendation, a request for improvement.

Currently, all exceptions are handled using the same class: **DocumentClientException**. In order to understand the true nature of an exception, when using try-catch, is necessary to check the Message string property.

For example, to detect an exception thrown when a database doesn't exist:
```cs
catch (DocumentClientException dce) when (dce.Message.Contains("Owner resource does not exist"))
```

To detect an exception thrown when a collection doesn't exist inside a database:
```cs
catch (DocumentClientException dce) when (dce.Message.Contains("Resource Not Found"))
```

To detect an exception thrown when a query has invalid syntax:
```cs
catch (DocumentClientException dce) when (dce.Message.Contains("Syntax error"))
```

Even worse, the dce.Message is used to store details serialized in JSON format, prefixed with a "Message: " string. So, if I want to get the root cause of a query syntax error, I need to make some dull string operations to:

- check for "Syntax Error" inside the exception message
- cut the "Message: " prefix of exception message
- deserialize the resulting string
- read the "Message" property inside the deserialized dynamic object

This is of course a bad practice, it's unclear why specialized exceptions weren't defined, removing the need for exotic error messages containing JSON.

It's unclear, for example, why this wasn't done:
```cs
public class DatabaseNotFoundException : DocumentClientException {
   ...
}

public class QuerySyntaxException : DocumentClientException {
   ...
}
```

Best Regards,
Roberto Prevato

**Comment by ayngling** [comment]
I'd like to second the post about stored procedures. If instead of throwing a string I throw an object (to include an error code) I just get {"Errors": ["Encountered exception while executing function. Exception = Error" Exception = [object Object]... Being able to at least get the contents of what was thrown (be is it a string or an object) from the DocumentClientException would be useful.

**Comment by admilazz** [comment]
I don't think we need specialized exception types, but we do need some kind of error code describing what happened. When interacting with particular underlying APIs, we should also get error codes matching the standard error codes for that API. (E.g. when interacting with the MongoDB API, there should be a field giving the MongoDB error code, in addition to any CosmosDB error code you might add.) That way applications coming to CosmosDB can maintain their error handling logic.

It's just no good to have to look for strings inside of the exception message.
